src/slack/app.rs

```diff
@@ -0,0 +1,214 @@
+                .header(http::header::LOCATION, "/e?identityAdded=error")
+                .finish(),
+        ));
+    }
+    let get = cis_client.clone();
+    let get_uid = user_id.user_id.clone();
+    Box::new(
+        Client::default()
+            .get(slack_token_url)
+            .header(http::header::USER_AGENT, "whoami")
+            .send()
+            .map_err(Into::into)
+            .and_then(move |mut res| res.json::<SlackTokenResponse>().map_err(Into::into))
+            .and_then(move |j| {
+                get.get_user_by(&get_uid, &GetBy::UserId, None)
+                    .and_then(move |profile: Profile| {
+                        update_slack(
+                            format!(
+                                "{}{}",
+                                slack_uri_data.direct_message_uri.to_string(),
+                                j.user.id
+                            ),
+                            j.user.name,
+                            profile,
+                            get.get_secret_store(),
+                        )
+                        .into_future()
+                        .map_err(Into::into)
+                    })
+                    .map_err(Into::into)
+            })
+            .and_then(move |profile: Profile| {
+                cis_client
+                    .update_user(&user_id.user_id, profile)
+                    .map_err(Into::into)
+            })
+            .and_then(|_| {
+                HttpResponse::Found()
+                    .header(http::header::LOCATION, "/e?identityAdded=slack")
+                    .finish()
+            }),
+    )
+}
+
+pub fn slack_app<T: AsyncCisClientTrait + 'static>(
+    slack: &Slack,
+    whoami: &WhoAmI,
+    secret: &[u8],
+    ttl_cache: Arc<RwLock<TtlCache<String, String>>>,
+    cis_client: T,
+) -> impl HttpServiceFactory {
+    let slack_client_id = ClientId::new(slack.client_id.clone());
+    let slack_client_secret = ClientSecret::new(slack.client_secret.clone());
+    let slack_auth_params = format!(
+        "?client_id={}&client_secret={}&redirect_uri={}",
+        &slack.client_id, &slack.client_secret, &slack.redirect_uri
+    );
+    let auth_url = AuthUrl::new(Url::parse(AUTH_URL).expect("Invalid authorization endpoint URL"));
+    let token_url = TokenUrl::new(
+        Url::parse(&format!("{}{}", TOKEN_URL, slack_auth_params))
+            .expect("Invalid token endpoint URL"),
+    );
+
+    let client = Arc::new(
+        BasicClient::new(
+            slack_client_id,
+            Some(slack_client_secret),
+            auth_url,
+            Some(token_url),
+        )
+        .add_scope(Scope::new(slack.scope.to_string()))
+        .set_redirect_url(RedirectUrl::new(
+            Url::parse(&slack.redirect_uri).expect("Invalid redirect URL"),
+        )),
+    );
+    let slack_uri_data: SlackUriData = SlackUriData {
+        slack_auth_params,
+        direct_message_uri: slack.direct_message_uri.to_string(),
```

Comment:
that should be a `.clone()` instead of `.to_string()` to make it more clear what's happening.